[#29 ] 관리자 기능 - 회원의 강제 탈퇴기능 / 전체 회원 조회 기능 구현

donsonioc2010 opened the pull request:
관리자기능으로, 지정한 사용자의 유저정보 삭제, 전체 회원정보에 대하여 조회할 수 있는 기능 제작

donsonioc2010 committed:
refactor : DTO의 오탈자 수정

donsonioc2010 committed:
feat : Forbidden에 대한 Exception 추가

donsonioc2010 committed:
feat : 관리자 권한의 전체 회원 조회기능, 회원의 삭제기능 구현

donsonioc2010 committed:
refactor : 계정정보 조회 Return해야할 Exception 수정

donsonioc2010 committed:
refactor : Exception Forbidden-> BadRequest로 발생하는 문제 수정

donsonioc2010 committed:
test : 테스트케이스 추가

donsonioc2010 committed:
Merge Develop Branch

donsonioc2010 commented on src/main/java/com/nooblol/community/service/impl/AdminServiceImpl.java:
```diff
@@ -51,10 +62,82 @@ public ResponseDto addAdminMember(
     return userSignOutService.signOutUser(userSignOutDto);
   }
 
+  /*
+    TODO: [22. 09. 10] : 해당 부분에서 초기 생각은 요청자의 정보를 받아 DB에서 관리자가 맞는지 확인을 한 이후 작업하는 로직이었슴.
+    다시 생각해보니 HttpSession에서 정보를 가져온 이후 대조하는게 더 맞는 판단으로 보이나,
+    현재 Login정보가 담긴 기능이 제작되지 않은 상황. 로그인기능의 수정을 진행하면서 해당 영역의 수정이 필요함.
+   */
```
해당 관리자 서비스의 경우 이전 리뷰에서 받은 내용이었던 관리자가 아니어도 제작이 가능하도록 작성되었던 로직을 생각해서 수정을 하고자 합니다.

Session을 활용하는 쪽으로 수정을 진행하고자 하나, 현재 Session에 로그인정보를 남기는 기능을 개발하지 않은 상황입니다. 

그래서 해당 부분에 대해서는   #37 에 대한 PR을 진행할 때 수정을 할 예정입니다.

donsonioc2010 committed:
docs : 기능의 주석 추가

f-lab-kyle commented on src/main/java/com/nooblol/community/controller/AdminController.java:
```diff
@@ -51,4 +53,28 @@ public ResponseDto signOutAdminMember(
+   * @param session
+   * @return
+   */
+  @PostMapping("/userList")
```
GET으로 하지 않는 이유가 궁금합니다.

donsonioc2010 replied:
아!

f-lab-kyle commented on src/main/java/com/nooblol/community/service/impl/AdminServiceImpl.java:
```diff
@@ -51,10 +62,82 @@ public ResponseDto addAdminMember(
+   */
+  @Override
+  public ResponseDto forceDeleteUser(AdminDeleteUserDto adminDeleteUserDto) {
+    try {
+      String adminPassword = EncryptUtils.stringChangeToSha512(
+          adminDeleteUserDto.getAdminUserPassword()
+      );
+      adminDeleteUserDto.setAdminUserPassword(adminPassword);
+
+      UserDto adminDbData = adminMapper.selectAdminDto(adminDeleteUserDto);
+
+      if (ObjectUtils.isEmpty(adminDbData) ||
+          adminDbData.getUserRole() != UserRoleStatus.ADMIN.getRoleValue()) {
+        throw new IllegalArgumentException(ExceptionMessage.FORBIDDEN);
+      }
+
+      ResponseDto resultDto = ResponseEnum.OK.getResponse();
+      resultDto.setResult(adminMapper.forcedDeleteUser(adminDeleteUserDto.getDeleteUserId()) > 0);
+
+      return resultDto;
+    } catch (NoSuchAlgorithmException | UnsupportedEncodingException e) {
+      throw new IllegalArgumentException(ExceptionMessage.SERVER_ERROR, e);
+    } catch (IllegalArgumentException e) {
+      //내부 발생이 Catch로 잡혀서 따로 분류
+      throw new IllegalArgumentException(ExceptionMessage.FORBIDDEN);
+    } catch (Exception e) {
+      throw new IllegalArgumentException(ExceptionMessage.BAD_REQUEST);
+    }
```
예외를 이렇게 전체로 잡기 보다 작은 단위로 예외를 잡아 처리하는게 좋을것 같단 생각이 드네요.

f-lab-kyle commented on src/main/resources/mybatis/mapper/community/AdminMapper.xml:
```diff
@@ -0,0 +1,37 @@
+  </select>
+
+  <select id="getAllUserList">
+    SELECt user_id,
+           user_email,
+           user_name,
+           user_role,
+           level,
+           exp,
+           created_at,
+           updated_at
+    FROM users
```
헉, 조건 없는 질의는 db가 힘들어 하지 않을까요? (왜 전부를 조회해야 하는지는 모르겠지만 그럴 상황이라면 페이징이라도 넣어야 하지 않을까요?)

donsonioc2010 replied:
아!

donsonioc2010 committed:
refactor : Review수정

- 삭제의 경우 DeleteMapping으로 수정
- 회원정보 조회의 경우 Get방식으로 수정
- 회원정보 조회시 Limit와 Page에 대한 파라미터 받도록 수정

donsonioc2010 committed:
test : 비즈니스 로직의 수정에 따른 테스트케이스 수 및 Builder추가

donsonioc2010 committed:
refactor : builder의 적용범위 수정
